### DISCUSSION TITLE
Node2Vec Question: how does the implementation generate input-target context pairs from the sampled random walks for training with the Skip-gram architecture?

### DISCUSSION THREAD

**Discussion by ghost** [pyg-team/pytorch_geometric#4524] [Q&A] [answered]
Hi, 

Thank you first of all for your great work! 

In your Pytorch Geometric Node2Vec tutorial you explain how the "context size" is used to generate smaller random walks from a larger random walk (see below for the example you give in the tutorial). 
![image](https://user-images.githubusercontent.com/76945330/164996790-067ffdf1-bc1d-4dd9-8324-e8715ff20808.png)

Now, how does your implementation generate input-target context pairs as training data for the Skip-gram architecture from this random walk? Which pairs in the first mini random walk {**u**: s4, s5, s6} will serve as training pairs from the algorithm? Will those be {u,s4}, {u,s5} and {u,s6} or {u,s4}, {s4,s5},{s5,s6}? 

Or, is the training data generated in the same way as for sentences that serve as input to Skip-gram in word2vec, where a sliding window is used over the random walk? (See image below):
![image](https://user-images.githubusercontent.com/76945330/165063887-752efbea-6a4d-40ef-bd9d-d8f222a3c5fa.png)

Thank you in advance! 
Ruth

**Comment by rusty1s** [comment, marked as answer]
Context nodes are generated via sliding window based on the original random walk (like in your first image). Training data for skipgram will then be {u, s4}, {u, s5}, ..., see [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/node2vec.py#L123).

**Reply by ghost** [reply]
Thank you!
